### PR TITLE
Some code cleanup in tensor mechanics

### PR DESCRIPTION
**Pull request by hchen139**
refs # 6810

### CODE REVIEW

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 1]
```diff
@@ -40,45 +35,27 @@ void
...
+  unsigned int _ndisp = displacements.size();
+  std::vector<VariableName> coupled_displacements;
+  for (unsigned int i = 0; i < _ndisp; ++i)
     coupled_displacements.push_back(displacements[i]);
 
+  std::vector<AuxVariableName> this_save_in = getParam<std::vector<AuxVariableName> >("save_in");
+  std::vector<std::vector<AuxVariableName> > save_in(_ndisp);
+  if (isParamValid("save_in"))
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      save_in[i].push_back(this_save_in[i]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: Wait this needs a check if enough save_in were supplied

Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: Looks like this or a similar one is crashing the test!

File: modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C [hunk 2]
```diff
@@ -35,24 +34,27 @@ void
 TensorMechanicsAxisymmetricRZAction::act()
 {
   std::vector<NonlinearVariableName> displacements = getParam<std::vector<NonlinearVariableName> > ("displacements");
-  std::vector<VariableName> coupled_displacements;
-  unsigned int dim = displacements.size();
+  unsigned int _ndisp = displacements.size();
 
   //Error checking:  Can only take two displacement variables in StressDivergenceRZTensors kernel
-  mooseAssert(dim == 2, "Expected two displacement variables but recieved " << dim);
+  mooseAssert(_ndisp == 2, "Expected two displacement variables but recieved " << _ndisp);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C]: Can you make this an actual conditional and a mooseerror?

File: modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C [hunk 3]
```diff
@@ -35,24 +34,27 @@ void
...
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      save_in[i].push_back(this_save_in[i]);
 
-  if (isParamValid("save_in_disp_z"))
-    save_in[1] = getParam<std::vector<AuxVariableName> >("save_in_disp_z");
+  std::vector<AuxVariableName> this_diag_save_in = getParam<std::vector<AuxVariableName> >("diag_save_in");
+  std::vector<std::vector<AuxVariableName> > diag_save_in(_ndisp);
+  if (isParamValid("diag_save_in"))
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      diag_save_in[i].push_back(this_diag_save_in[i]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C]: Size of diag_save_in needs to be checked

**Comment by dschwen** [comment]
@hchen139 the app test failures in Marmot are unrelated to this PR, the failure in Grizzly however is a result of this. I suggest this here gets merged and then _somebody_ cleans up the Grizzly files to remove the deprecated `save_in_disp_x` etc . and replaces them with the new syntax.

**Comment by permcody** [comment]
@hchen139 - do you have access to Grizzly?

**Comment by hchen139** [comment]
@permcody No, it is not even in my security list. We talked with @bwspenc about this, I think he would like to clean up the Grizzly files.

**Comment by bwspenc** [comment]
I've updated the grizzly tests, so I can submit a grizzly patch as soon as this gets merged to devel.

**Comment by permcody** [comment]
We'll take care of this next week.

**Comment by bwspenc** [comment]
OK.  Just let me know when you're ready to do it so I can be ready with my commit.

**Comment by hchen139** [comment]
@permcody @dschwen @bwspenc The latest commit is compatible with both input file syntaxes

**Comment by permcody** [comment]
@hchen139 - You should probably use https://github.com/idaholab/moose/blob/devel/framework/include/utils/InputParameters.h#L206-L210

This saves you from having to check if parameters are valid then using MooseDeprecated. Also these parameters work for parameters that have default values where your `isValid()` checks would not.

**Comment by dschwen** [comment]
I do not understand why we introduce _new_ parameters that are using the same schema as the parameters we have been trying to _deprecate_ for months now. I'd much rather see the app tests fixed and _only_ the vector form of the `save_in`, rather than introducing parameters that are about to be removed anyways.

**Comment by hchen139** [comment]
will do that @permcody

**Comment by permcody** [comment]
This looks OK to me, but I'll wait for @dschwen to return to make sure this is the correct direction.

**Comment by permcody** [comment]
@dschwen - please review

**Comment by dschwen** [comment]
Wait, @permcody and @hchen139 is this ready for a review?

**Comment by permcody** [comment]
I believe it is from @hchen139's point of view. You had indicated that you didn't like the save_in direction so I felt that we needed to discuss a bit more.

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 4]
```diff
@@ -46,17 +36,6 @@ TensorMechanicsAction::act()
   std::vector<NonlinearVariableName> displacements;
   if (isParamValid("displacements"))
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: Just make the parameter required. No need for custom errors and checking

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 5]
```diff
@@ -40,45 +34,35 @@ void
 TensorMechanicsAction::act()
 {
   std::vector<NonlinearVariableName> displacements;
-  //This code will be the up-to-date version using the displacement string
   if (isParamValid("displacements"))
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: Just make the parameter required. No need for custom checks and error messages.

File: modules/tensor_mechanics/src/actions/TensorMechanicsAction.C [hunk 6]
```diff
@@ -40,45 +34,35 @@ void
...
-    save_in[2] = getParam<std::vector<AuxVariableName> >("save_in_disp_z");
-
+  std::vector<std::vector<AuxVariableName> > diag_save_in(_ndisp);
+  if (isParamValid("diag_save_in"))
+  {
+    std::vector<AuxVariableName> this_diag_save_in = getParam<std::vector<AuxVariableName> >("diag_save_in");
+    if (this_diag_save_in.size() != _ndisp)
+      mooseError("Number of diag_save_in variables should equal to the number of displacement variables " << _ndisp);
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      diag_save_in[i].push_back(this_diag_save_in[i]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAction.C]: This logic seems overly complicated. Just get the parameter directly into `diag_save_in` and do the size check on that variable.

**Comment by hchen139** [comment]
@dschwen @permcody this commit removes all the old syntaxes in solid mechanics for displacements, save_ins and diag_save_ins. It will definitely break some tests in Grizzly. And @bwspenc will update the corresponding tests. Thanks.

File: modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C [hunk 7]
```diff
@@ -35,24 +33,36 @@ void
...
 
   //Error checking:  Can only take two displacement variables in StressDivergenceRZTensors kernel
-  mooseAssert(dim == 2, "Expected two displacement variables but recieved " << dim);
+  if (_ndisp != 2)
+    mooseError("Number of displacement variables should be 2 but recieved " << _ndisp);
 
-  for (unsigned int i = 0; i < dim; ++i)
+  std::vector<VariableName> coupled_displacements;
+  for (unsigned int i = 0; i < _ndisp; ++i)
     coupled_displacements.push_back(displacements[i]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C]: What is this copy for?

File: modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C [hunk 8]
```diff
@@ -35,24 +33,36 @@ void
...
-  if (isParamValid("save_in_disp_z"))
-    save_in[1] = getParam<std::vector<AuxVariableName> >("save_in_disp_z");
+  std::vector<std::vector<AuxVariableName> > diag_save_in(_ndisp);
+  if (isParamValid("diag_save_in"))
+  {
+    std::vector<AuxVariableName> this_diag_save_in = getParam<std::vector<AuxVariableName> >("diag_save_in");
+    if (this_diag_save_in.size() != _ndisp)
+      mooseError("Number of diag_save_in variables should equal to the number of displacement variables " << _ndisp);
+    for (unsigned int i = 0; i < _ndisp; ++i)
+      diag_save_in[i].push_back(this_diag_save_in[i]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsAxisymmetricRZAction.C]: again, get rid of `this_diag_save_in ` and `this_save_in `

File: modules/tensor_mechanics/src/actions/TensorMechanicsRSphericalAction.C [hunk 9]
```diff
@@ -33,19 +34,36 @@ void
...
 
   // Error checking:  Can only take one displacement variable in StressDivergenceRSphericalTensors kernel
-  mooseAssert(dim == 1, "Expected a single displacement variable but recieved " << dim);
+  if (_ndisp != 1)
+    mooseError("Number of displacement variable should be 1 but recieved " << _ndisp);
 
-  for (unsigned int i = 0; i < dim; ++i)
+  std::vector<VariableName> coupled_displacements;
+  for (unsigned int i = 0; i < _ndisp; ++i)
     coupled_displacements.push_back(displacements[i]);
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/TensorMechanicsRSphericalAction.C]: get rid of `coupled_displacements` here, too

**Comment by dschwen** [comment]
Yes, this definitely goes in the right direction. Per my comments I think this can be simplified a little bit. Once that is done it should be ready to merge.

**Comment by dschwen** [comment]
Looks good to me now!

**Comment by permcody** [comment]
@hchen139 - Please let me know when you have your application patch ready. We can merge this into devel once you prepared to patch Grizzly.

**Comment by bwspenc** [comment]
@permcody I have a Grizzly patch ready.  Just let me know when you're ready to merge this, and I'll update the moose submodule in my patch.

**Comment by permcody** [comment]
👍

**Comment by permcody** [comment]
Merged - please update Grizzly

**Comment by bwspenc** [comment]
OK, working on it right now.

**Pull request merged**